File: src/main/java/com/projectswg/holocore/resources/support/npc/spawn/Spawner.java
/***********************************************************************************
 * Copyright (c) 2018 /// Project SWG /// www.projectswg.com                       *
 *                                                                                 *
 * ProjectSWG is the first NGE emulator for Star Wars Galaxies founded on          *
 * July 7th, 2011 after SOE announced the official shutdown of Star Wars Galaxies. *
 * Our goal is to create an emulator which will provide a server for players to    *
 * continue playing a game similar to the one they used to play. We are basing     *
 * it on the final publish of the game prior to end-game events.                   *
 *                                                                                 *
 * This file is part of Holocore.                                                  *
 *                                                                                 *
 * --------------------------------------------------------------------------------*
 *                                                                                 *
 * Holocore is free software: you can redistribute it and/or modify                *
 * it under the terms of the GNU Affero General Public License as                  *
 * published by the Free Software Foundation, either version 3 of the              *
 * License, or (at your option) any later version.                                 *
 *                                                                                 *
 * Holocore is distributed in the hope that it will be useful,                     *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of                  *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                   *
 * GNU Affero General Public License for more details.                             *
 *                                                                                 *
 * You should have received a copy of the GNU Affero General Public License        *
 * along with Holocore.  If not, see <http://www.gnu.org/licenses/>.               *
 ***********************************************************************************/
package com.projectswg.holocore.resources.support.npc.spawn;

import com.projectswg.common.data.encodables.tangible.PvpFaction;
import com.projectswg.common.data.location.Location;
import com.projectswg.holocore.resources.support.data.server_info.loader.DataLoader;
import com.projectswg.holocore.resources.support.data.server_info.loader.NpcLoader.*;
import com.projectswg.holocore.resources.support.data.server_info.loader.NpcPatrolRouteLoader.PatrolRouteWaypoint;
import com.projectswg.holocore.resources.support.data.server_info.loader.NpcPatrolRouteLoader.PatrolType;
import com.projectswg.holocore.resources.support.data.server_info.loader.NpcStatLoader.DetailNpcStatInfo;
import com.projectswg.holocore.resources.support.data.server_info.loader.NpcStatLoader.NpcStatInfo;
import com.projectswg.holocore.resources.support.data.server_info.loader.NpcStaticSpawnLoader.PatrolFormation;
import com.projectswg.holocore.resources.support.data.server_info.loader.NpcStaticSpawnLoader.StaticSpawnInfo;
import com.projectswg.holocore.resources.support.objects.swg.SWGObject;
import com.projectswg.holocore.resources.support.objects.swg.building.BuildingObject;
import com.projectswg.holocore.resources.support.objects.swg.creature.CreatureDifficulty;
import com.projectswg.holocore.resources.support.objects.swg.custom.AIBehavior;
import com.projectswg.holocore.services.support.objects.ObjectStorageService.BuildingLookup;
import me.joshlarson.jlcommon.log.Log;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.Objects;
import java.util.Random;
import java.util.stream.Collectors;

public final class Spawner {
	
	private final StaticSpawnInfo spawn;
	private final NpcInfo npc;
	private final NpcStatInfo npcStat;
	private final DetailNpcStatInfo npcDetailStat;
	
	private final Location location;
	private final List<ResolvedPatrolWaypoint> waypoints;
	private final SWGObject egg;
	private final Random random;
	
	public Spawner(@NotNull StaticSpawnInfo spawn, @NotNull SWGObject egg) {
		this.spawn = Objects.requireNonNull(spawn, "spawn");
		this.npc = DataLoader.npcs().getNpc(spawn.getNpcId());
		Objects.requireNonNull(npc, "Invalid npc id: " + spawn.getNpcId());
		this.npcStat = DataLoader.npcStats().getNpcStats(npc.getCombatLevel());
		Objects.requireNonNull(npcStat, "Invalid npc combat lebel: " + npc.getCombatLevel());
		
		this.location = Location.builder()
					.setTerrain(spawn.getTerrain())
					.setPosition(spawn.getX(), spawn.getY(), spawn.getZ())
					.setHeading(spawn.getHeading())
					.build();
		if (spawn.getPatrolId() < 1000) {
			this.waypoints = null;
		} else {
			List<PatrolRouteWaypoint> waypoints = Objects.requireNonNull(DataLoader.npcPatrolRoutes().getPatrolRoute(spawn.getPatrolId()), "Invalid patrol route: " + spawn.getPatrolId());
			this.waypoints = waypoints.stream().map(ResolvedPatrolWaypoint::new).collect(Collectors.toList());
		}
		this.egg = Objects.requireNonNull(egg, "egg");
		this.random = new Random();
		
		switch (npc.getDifficulty()) {
			case NORMAL:
			default:
				this.npcDetailStat = npcStat.getNormalDetailStat();
				break;
			case ELITE:
				this.npcDetailStat = npcStat.getEliteDetailStat();
				break;
			case BOSS:
				this.npcDetailStat = npcStat.getBossDetailStat();
				break;
		}
	}
	
	/**
	 * Calculates a random number between {@code minRespawnDelay} and
	 * {@code maxRespawnDelay}
	 * @return a random number between {@code minRespawnDelay} and
	 * {@code maxRespawnDelay}
	 */
	public int getRespawnDelay() {
		return random.nextInt((getMaxSpawnTime() - getMinSpawnTime()) + 1) + getMinSpawnTime();
	}
	
	/**
	 * @return a random IFF template
	 */
	public String getRandomIffTemplate() {
		return getRandom(getIffs());
	}
	
	public String getRandomPrimaryWeapon() {
		return getRandom(getPrimaryWeapons());
	}
	
	public String getRandomSecondaryWeapon() {
		return getRandom(getSecondaryWeapons());
	}
	
	public List<ResolvedPatrolWaypoint> getPatrolRoute() {
		return waypoints;
	}
	
	public SWGObject getEgg() {
		return egg;
	}
	
	public Location getLocation() {
		return location;
	}
	
	public int getId() {
		return spawn.getId();
	}
	
	public String getSpawnerType() {
		return spawn.getSpawnerType();
	}
	
	public String getNpcId() {
		return spawn.getNpcId();
	}
	
	public String getBuildingId() {
		return spawn.getBuildingId();
	}
	
	public String getMood() {
		return spawn.getMood();
	}
	
	public AIBehavior getBehavior() {
		return spawn.getBehavior();
	}
	
	public int getPatrolId() {
		return spawn.getPatrolId();
	}
	
	public PatrolFormation getPatrolFormation() {
		return spawn.getPatrolFormation();
	}
	
	public int getLoiterRadius() {
		return spawn.getLoiterRadius();
	}
	
	public int getMinSpawnTime() {
		return spawn.getMinSpawnTime();
	}
	
	public int getMaxSpawnTime() {
		return spawn.getMaxSpawnTime();
	}
	
	public int getAmount() {
		return spawn.getAmount();
	}
	
	public SpawnerFlag getSpawnerFlag() {
		return npc.getSpawnerFlag();
	}
	
	public CreatureDifficulty getDifficulty() {
		return npc.getDifficulty();
	}
	
	public int getCombatLevel() {
		return npc.getCombatLevel();
	}
	
	public String getName() {
		return npc.getName();
	}
	
	public String getStfName() {
		return npc.getStfName();
	}
	
	public List<String> getIffs() {
		return npc.getIffs();
	}
	
	public PvpFaction getFaction() {
		return npc.getFaction();
	}
	
	public boolean isSpecForce() {
		return npc.isSpecForce();
	}
	
	public double getAttackSpeed() {
		return npc.getAttackSpeed();
	}
	
	public double getMovementSpeed() {
		return npc.getMovementSpeed();
	}
	
	public List<String> getPrimaryWeapons() {
		return npc.getPrimaryWeapons().stream().map(DataLoader.npcWeapons()::getWeapons).filter(Objects::nonNull).flatMap(List::stream).collect(Collectors.toList());
	}
	
	public List<String> getSecondaryWeapons() {
		return npc.getSecondaryWeapons().stream().map(DataLoader.npcWeapons()::getWeapons).filter(Objects::nonNull).flatMap(List::stream).collect(Collectors.toList());
	}
	
	public double getPrimaryWeaponSpeed() {
		return npc.getPrimaryWeaponSpeed();
	}
	
	public double getSecondaryWeaponSpeed() {
		return npc.getSecondaryWeaponSpeed();
	}
	
	public int getAggressiveRadius() {
		return npc.getAggressiveRadius();
	}
	
	public int getAssistRadius() {
		return npc.getAssistRadius();
	}
	
	public boolean isDeathblow() {
		return npc.isDeathblow();
	}
	
	public String getLootTable1() {
		return npc.getLootTable1();
	}
	
	public String getLootTable2() {
		return npc.getLootTable2();
	}
	
	public String getLootTable3() {
		return npc.getLootTable3();
	}
	
	public int getLootTable1Chance() {
		return npc.getLootTable1Chance();
	}
	
	public int getLootTable2Chance() {
		return npc.getLootTable2Chance();
	}
	
	public int getLootTable3Chance() {
		return npc.getLootTable3Chance();
	}
	
	public HumanoidNpcInfo getHumanoidInfo() {
		return npc.getHumanoidInfo();
	}
	
	public DroidNpcInfo getDroidInfo() {
		return npc.getDroidInfo();
	}
	
	public CreatureNpcInfo getCreatureInfo() {
		return npc.getCreatureInfo();
	}
	
	public int getLevel() {
		return npcStat.getLevel();
	}
	
	public int getHealthRegen() {
		return npcStat.getHealthRegen();
	}
	
	public int getActionRegen() {
		return npcStat.getActionRegen();
	}
	
	public int getMindRegen() {
		return npcStat.getMindRegen();
	}
	
	public int getHealth() {
		return npcDetailStat.getHealth();
	}
	
	public int getAction() {
		return npcDetailStat.getAction();
	}
	
	public int getRegen() {
		return npcDetailStat.getRegen();
	}
	
	public int getCombatRegen() {
		return npcDetailStat.getCombatRegen();
	}
	
	public int getDamagePerSecond() {
		return npcDetailStat.getDamagePerSecond();
	}
	
	public int getToHit() {
		return npcDetailStat.getToHit();
	}
	
	public int getDef() {
		return npcDetailStat.getDef();
	}
	
	public int getArmor() {
		return npcDetailStat.getArmor();
	}
	
	public int getXp() {
		return npcDetailStat.getXp();
	}
	
	private <T> T getRandom(List<T> list) {
		return list.get(random.nextInt(list.size()));
	}
	
	public static class ResolvedPatrolWaypoint {
		
		private final SWGObject parent;
		private final Location location;
		private final double delay;
		private final PatrolType patrolType;
		
		private ResolvedPatrolWaypoint(PatrolRouteWaypoint waypoint) {
			this.parent = getPatrolWaypointParent(waypoint);
			this.location = getPatrolWaypointLocation(waypoint);
			this.delay = waypoint.getDelay();
			this.patrolType = waypoint.getPatrolType();
		}
		
		public SWGObject getParent() {
			return parent;
		}
		
		public Location getLocation() {
			return location;
		}
		
		public double getDelay() {
			return delay;
		}
		
		public PatrolType getPatrolType() {
			return patrolType;
		}
		
		private static Location getPatrolWaypointLocation(PatrolRouteWaypoint waypoint) {
			return Location.builder()
					.setTerrain(waypoint.getTerrain())
					.setX(waypoint.getX())
					.setY(waypoint.getY())
					.setZ(waypoint.getZ()).build();
		}
		
		private static SWGObject getPatrolWaypointParent(PatrolRouteWaypoint waypoint) {
			if (waypoint.getBuildingId().isEmpty() || waypoint.getBuildingId().endsWith("_world"))
				return null;
			
			BuildingObject building = BuildingLookup.getBuildingByTag(waypoint.getBuildingId());
			if (building == null) {
				Log.w("PatrolRouteWaypoint: Invalid building id for patrol id: %d and group id: %d", waypoint.getPatrolId(), waypoint.getGroupId());
				return null;
			}
			
			SWGObject cell = building.getCellByNumber(waypoint.getCellId());
			if (cell == null) {
				Log.w("PatrolRouteWaypoint: Invalid cell [%d] for building: %s, patrol id: %d and group id: %d", waypoint.getCellId(), waypoint.getBuildingId(), waypoint.getPatrolId(), waypoint.getGroupId());
				return null;
			}
			
			return cell;
		}
		
	}
	
}

File: src/main/java/com/projectswg/holocore/resources/support/npc/spawn/Spawner.kt
/***********************************************************************************
 * Copyright (c) 2018 /// Project SWG /// www.projectswg.com                       *
 * *
 * ProjectSWG is the first NGE emulator for Star Wars Galaxies founded on          *
 * July 7th, 2011 after SOE announced the official shutdown of Star Wars Galaxies. *
 * Our goal is to create an emulator which will provide a server for players to    *
 * continue playing a game similar to the one they used to play. We are basing     *
 * it on the final publish of the game prior to end-game events.                   *
 * *
 * This file is part of Holocore.                                                  *
 * *
 * --------------------------------------------------------------------------------*
 * *
 * Holocore is free software: you can redistribute it and/or modify                *
 * it under the terms of the GNU Affero General Public License as                  *
 * published by the Free Software Foundation, either version 3 of the              *
 * License, or (at your option) any later version.                                 *
 * *
 * Holocore is distributed in the hope that it will be useful,                     *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of                  *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                   *
 * GNU Affero General Public License for more details.                             *
 * *
 * You should have received a copy of the GNU Affero General Public License        *
 * along with Holocore.  If not, see <http:></http:>//www.gnu.org/licenses/>.               *
 */
package com.projectswg.holocore.resources.support.npc.spawn

import com.projectswg.common.data.location.Location
import com.projectswg.holocore.resources.support.data.server_info.loader.DataLoader
import com.projectswg.holocore.resources.support.data.server_info.loader.combat.FactionLoader.Faction
import com.projectswg.holocore.resources.support.data.server_info.loader.npc.NpcLoader.*
import com.projectswg.holocore.resources.support.data.server_info.loader.npc.NpcPatrolRouteLoader.PatrolRouteWaypoint
import com.projectswg.holocore.resources.support.data.server_info.loader.npc.NpcPatrolRouteLoader.PatrolType
import com.projectswg.holocore.resources.support.data.server_info.loader.npc.NpcStaticSpawnLoader.*
import com.projectswg.holocore.resources.support.objects.swg.SWGObject
import com.projectswg.holocore.resources.support.objects.swg.creature.CreatureDifficulty
import com.projectswg.holocore.resources.support.objects.swg.custom.AIBehavior
import com.projectswg.holocore.resources.support.objects.swg.custom.AIObject
import com.projectswg.holocore.services.support.objects.ObjectStorageService.BuildingLookup
import me.joshlarson.jlcommon.log.Log
import java.util.*
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.ThreadLocalRandom
import java.util.stream.Collectors

class Spawner(spawn: StaticSpawnInfo, egg: SWGObject) {
	
	private val spawn: StaticSpawnInfo = Objects.requireNonNull(spawn, "spawn")
	private val npc: NpcInfo = Objects.requireNonNull(DataLoader.npcs().getNpc(spawn.npcId), "Invalid npc id: " + spawn.npcId)
	private val npcsInternal = CopyOnWriteArrayList<AIObject>()
	
	val location: Location = Location.builder().setTerrain(spawn.terrain).setPosition(spawn.x, spawn.y, spawn.z).setHeading(spawn.heading.toDouble()).build()
	val patrolRoute: List<ResolvedPatrolWaypoint>?
	val egg: SWGObject = Objects.requireNonNull(egg, "egg")
	val npcs: List<AIObject> = Collections.unmodifiableList(npcsInternal)
	
	/**
	 * Calculates a random number between `minRespawnDelay` and
	 * `maxRespawnDelay`
	 * @return a random number between `minRespawnDelay` and
	 * `maxRespawnDelay`
	 */
	val respawnDelay: Int
		get() = ThreadLocalRandom.current().nextInt(maxSpawnTime - minSpawnTime + 1) + minSpawnTime
	
	/**
	 * @return a random IFF template
	 */
	fun getRandomIffTemplate(): String = getRandom(iffs)
	fun getRandomPrimaryWeapon(): String = getRandom(primaryWeapons)
	fun getRandomSecondaryWeapon(): String = getRandom(secondaryWeapons)
	
	fun addNPC(obj: AIObject) {
		npcsInternal.add(obj)
	}
	
	fun removeNPC(obj: AIObject) {
		npcsInternal.remove(obj)
	}
	
	val id: String
		get() = spawn.id
	
	val spawnerType: String
		get() = spawn.spawnerType
	
	val npcId: String
		get() = spawn.npcId
	
	val buildingId: String
		get() = spawn.buildingId
	
	val mood: String
		get() = spawn.mood
	
	val behavior: AIBehavior
		get() = spawn.behavior
	
	val patrolId: String
		get() = spawn.patrolId
	
	val patrolFormation: PatrolFormation
		get() = spawn.patrolFormation
	
	val loiterRadius: Int
		get() = spawn.loiterRadius
	
	val minSpawnTime: Int
		get() = spawn.minSpawnTime
	
	val maxSpawnTime: Int
		get() = spawn.maxSpawnTime
	
	val amount: Int
		get() = spawn.amount
	
	val spawnerFlag: SpawnerFlag
		get() = spawn.spawnerFlag
	
	val difficulty: CreatureDifficulty
		get() = spawn.difficulty
	
	val minLevel: Int
		get() = spawn.minLevel
	
	val maxLevel: Int
		get() = spawn.maxLevel
	
	val name: String
		get() = npc.name
	
	val stfName: String
		get() = npc.stfName
	
	val iffs: List<String>
		get() = npc.iffs
	
	val faction: Faction?
		get() = npc.faction
	
	val isSpecForce: Boolean
		get() = npc.isSpecForce
	
	val attackSpeed: Double
		get() = npc.attackSpeed
	
	val movementSpeed: Double
		get() = npc.movementSpeed
	
	val scaleMin: Double
		get() = npc.scaleMin
	
	val scaleMax: Double
		get() = npc.scaleMax
	
	val hue: Int
		get() = npc.hue
	
	val primaryWeapons: List<String>
		get() = npc.primaryWeapons.stream().map { DataLoader.npcWeapons().getWeapons(it) }.filter { Objects.nonNull(it) }.flatMap { it.stream() }.collect(Collectors.toList())
	
	val secondaryWeapons: List<String>
		get() = npc.secondaryWeapons.stream().map { DataLoader.npcWeapons().getWeapons(it) }.filter { Objects.nonNull(it) }.flatMap { it.stream() }.collect(Collectors.toList())
	
	val aggressiveRadius: Int
		get() = npc.aggressiveRadius
	
	val assistRadius: Int
		get() = npc.assistRadius
	
	val isDeathblow: Boolean
		get() = npc.isDeathblow
	
	val lootTable1: String
		get() = npc.lootTable1
	
	val lootTable2: String
		get() = npc.lootTable2
	
	val lootTable3: String
		get() = npc.lootTable3
	
	val lootTable1Chance: Int
		get() = npc.lootTable1Chance
	
	val lootTable2Chance: Int
		get() = npc.lootTable2Chance
	
	val lootTable3Chance: Int
		get() = npc.lootTable3Chance
	
	val humanoidInfo: HumanoidNpcInfo
		get() = npc.humanoidInfo
	
	val droidInfo: DroidNpcInfo
		get() = npc.droidInfo
	
	val creatureInfo: CreatureNpcInfo
		get() = npc.creatureInfo
	
	init {
		if (spawn.patrolId.isEmpty() || spawn.patrolId == "0") { // TODO: Replace the latter with empty string
			this.patrolRoute = null
		} else {
			val waypoints = Objects.requireNonNull(DataLoader.npcPatrolRoutes().getPatrolRoute(spawn.patrolId), "Invalid patrol route: " + spawn.patrolId)
			this.patrolRoute = waypoints.stream().map { ResolvedPatrolWaypoint(it) }.collect(Collectors.toList())
		}
	}
	
	private fun <T> getRandom(list: List<T>): T {
		return list[ThreadLocalRandom.current().nextInt(list.size)]
	}
	
	class ResolvedPatrolWaypoint(private val waypoint: PatrolRouteWaypoint) {
		val parent: SWGObject?
		val location: Location
		
		val delay: Double
			get() = waypoint.delay
		
		val patrolType: PatrolType
			get() = waypoint.patrolType
		
		val groupId: String
			get() = waypoint.groupId
		
		val patrolId: String
			get() = waypoint.patrolId
		
		init {
			this.parent = getPatrolWaypointParent(waypoint)
			this.location = getPatrolWaypointLocation(waypoint)
		}
		
		private fun getPatrolWaypointLocation(waypoint: PatrolRouteWaypoint): Location {
			return Location.builder()
					.setTerrain(waypoint.terrain)
					.setX(waypoint.x)
					.setY(waypoint.y)
					.setZ(waypoint.z).build()
		}
		
		private fun getPatrolWaypointParent(waypoint: PatrolRouteWaypoint): SWGObject? {
			if (waypoint.buildingId.isEmpty() || waypoint.buildingId.endsWith("_world"))
				return null
			
			val building = BuildingLookup.getBuildingByTag(waypoint.buildingId)
			if (building == null) {
				Log.w("PatrolRouteWaypoint: Invalid building id for patrol id: %d and group id: %d", waypoint.patrolId, waypoint.groupId)
				return null
			}
			
			val cell = building.getCellByNumber(waypoint.cellId)
			if (cell == null) {
				Log.w("PatrolRouteWaypoint: Invalid cell [%d] for building: %s, patrol id: %d and group id: %d", waypoint.cellId, waypoint.buildingId, waypoint.patrolId, waypoint.groupId)
				return null
			}
			
			return cell
		}
		
	}
	
}
